Support data persistence: integrate with storageClass

gdasson opened the pull request:
This PR adds support for data persistence using StorageClasses. 

Supports setting VolumeClaimTemplate for the statefulset.

Fixes issue: https://github.com/etcd-io/etcd-operator/issues/28

ahrtr commented:
/ok-to-test

ahrtr commented on config/crd/bases/operator.etcd.io_etcdclusters.yaml:
```diff
@@ -42,13 +42,27 @@ spec:
+                x-kubernetes-int-or-string: true
             required:
             - size
+            - storageClassName
```
It isn't required. The default one will be used if not specified.

gdasson commented:
@ahrtr : I have fixed your review comments. Please review. 

Also, for now, I have removed the e2e tests that I had added(used gingko framework) in my previous commits. I'll add some e2e tests related to storage, probably in a follow-up PR, once the e2e test PR is merged.

ahrtr commented on internal/controller/utils.go:
```diff
@@ -142,6 +148,97 @@ func createOrPatchStatefulSet(ctx context.Context, logger logr.Logger, ec *ecv1a
+
+		switch ec.Spec.StorageSpec.AccessModes {
+		case corev1.ReadWriteOnce, "":
+			if ec.Spec.StorageSpec.StorageClassName == "" {
+				stsSpec.VolumeClaimTemplates = []corev1.PersistentVolumeClaim{
+					{
+						ObjectMeta: pvcObjectMeta,
+						Spec: corev1.PersistentVolumeClaimSpec{
+							AccessModes: []corev1.PersistentVolumeAccessMode{ec.Spec.StorageSpec.AccessModes},
+							Resources:   pvcResources,
+							// No StorageClassName defined. Use default storage class.
+						},
+					},
+				}
+			} else {
+				stsSpec.VolumeClaimTemplates = []corev1.PersistentVolumeClaim{
+					{
+						ObjectMeta: pvcObjectMeta,
+						Spec: corev1.PersistentVolumeClaimSpec{
+							AccessModes:      []corev1.PersistentVolumeAccessMode{ec.Spec.StorageSpec.AccessModes},
+							StorageClassName: &ec.Spec.StorageSpec.StorageClassName,
+							Resources:        pvcResources,
+						},
+					},
+				}
+			}
```
```suggestion			
			stsSpec.VolumeClaimTemplates = []corev1.PersistentVolumeClaim{
				{
					ObjectMeta: pvcObjectMeta,
					Spec: corev1.PersistentVolumeClaimSpec{
						AccessModes: []corev1.PersistentVolumeAccessMode{ec.Spec.StorageSpec.AccessModes},
						Resources:   pvcResources,
						// No StorageClassName defined. Use default storage class.
					},
				},
			}
			if ec.Spec.StorageSpec.StorageClassName != "" {
				stsSpec.VolumeClaimTemplates[0].Spec.StorageClassName = &ec.Spec.StorageSpec.StorageClassName
			}			
```

ahrtr commented:
> once the e2e test PR is merged.

The e2e PR has already been merged. But it's OK to add the e2e test separately. 

cc @hakman @ivanvc @ArkaSaha30

ahrtr commented:
@gdasson have you verified this PR against any storage?

gdasson commented:
@ahrtr : Yes, I tested this against the local path provisioner supported StorageClass that comes default with kind and Rancher Desktop clusters. That's what I had referenced in my e2e tests in previous commits. I removed those e2e tests because they were Gingko style.

ahrtr commented:
@gdasson Please add e2e test in followup PRs, thx
